Implemented HTML minifier module

leekelleher opened the pull request:
Implements #12 Create a HTML minifier module.

Adds a dependency on [WebMarkupMin.Core](https://www.nuget.org/packages/WebMarkupMin.Core/2.0.0-rc2) library, (source code is available on GitHub: [WebMarkupMin](https://github.com/Taritsyn/WebMarkupMin)).

Example usage:

```
Pipelines.Add("Content",
    ReadFiles("*.md"),
    FrontMatter(Yaml()),
    Markdown(),
    Razor(),
    HtmlMinify(),
    WriteFiles(".html")
);
```

There are many minification settings, which are all outlined on the [WebMarkupMin wiki](https://github.com/Taritsyn/WebMarkupMin/wiki/HTML-Minifier).  I've added fluent methods for these, as well as XML code comments.

So far I've added one unit-test.  I will need your input here, as TDD isn't my strongest skill; I found myself writing tests that were pretty much re-testing the WebMarkupMin library, (which already has its own set of  [`HtmlMinifier` unit-tests](https://github.com/Taritsyn/WebMarkupMin/blob/master/test/WebMarkupMin.Core.Test/HtmlMinifierTests.cs)).  I wasn't sure what else to test? (Do you want 100% coverage of the fluent methods?)

---

In terms of the minifcation functionality itself, I guess we should have a discussion around the default `HtmlMinificationSettings` settings, most of these are common-sense, but others feel a little more extreme, (such as `RemoveOptionalEndTags` is enabled by default).

I'm happy to discuss and revise the PR where needed.

leekelleher committed:
Implemented HTML minifier module

Uses "WebMarkupMin" library
Implements #12

miere43 commented:
Looks awesome. Can't wait to use it instead of additional dependency on Node.js.

daveaglick commented:
:thumbsup: This looks great! I'll take a closer look later today, but off the top of my head, I don't have any revisions on the code. Very clean.

In regards to exposing the other minification settings, I don't think we necessarily need to wrap all the possible behaviors. For modules that use more complicated libraries, I've taken the approach of exposing the most common or helpful options through the module and then provided a fluent method that exposes the underlying library in a delegate for more advanced configuration. For example, the `GitHub` module works like this.

As far as testing, I think what you've got works fine. In general, modules that use functionality from Wyam.Core need to be tested a little more (I.e., `ReadFiles`). For ones like this where we're mainly just testing pass-through and output to an underlying library, a single test like yours should do the trick.

leekelleher commented:
Cool, thanks @daveaglick! I'll look to prune the fluent methods back :sunglasses:

leekelleher committed:
HtmlMinify module: Removed many fluent methods

Replaced with a `WithSettings` fluent method. Users can now customise the minification settings.
Added supporting unit-test.

leekelleher commented:
Updated the PR. Added a `WithSettings` fluent method, replacing the majority of the previous options.

Do shout if there's an option you'd like as a fluent method, (_that's open to anyone reading this_).

daveaglick commented:
Thanks again for this - really fantastic!
